include/world_builder/point.h

```diff
@@ -76,59 +96,133 @@ namespace WorldBuilder
       /**
        * Destructor
        */
-      ~Point();
+      inline
+      ~Point() = default;
 
-      Point<dim> &operator=(const Point<dim> &point);
+      inline
+      Point<dim> &operator=(const Point<dim> &point_right)
+      {
+        point = point_right.point;
+        coordinate_system = point_right.coordinate_system;
+        return *this;
+      }
 
       /**
        * dot product
        */
-      double operator*(const Point<dim> &point) const;
+      inline
+      double operator*(const Point<dim> &point_right) const
+      {
+        const std::array<double,dim> &array = point_right.get_array();
+        double dot_product = 0;
+        for (unsigned int i = 0; i < dim; ++i)
+          dot_product += point[i] * array[i];
+        return dot_product;
+      }
 
 
       /**
        * Multiply the vector with a scalar
        */
-      Point<dim> operator*(const double scalar) const;
+      inline
+      Point<dim> operator*(const double scalar) const
+      {
+        // initialize the array to zero.
+        std::array<double,dim> array;
+        for (unsigned int i = 0; i < dim; ++i)
+          array[i] = point[i] * scalar;
+        return Point<dim>(array,coordinate_system);
+      }
 
       /**
        * Divide the vector through a scalar
        */
-      Point<dim> operator/(const double scalar) const;
+      inline
+      Point<dim> operator/(const double scalar) const
+      {
+        // initialize the array to zero.
+        std::array<double,dim> array;
+        const double one_over_scalar = 1/scalar;
+        for (unsigned int i = 0; i < dim; ++i)
+          array[i] = point[i] * one_over_scalar;
+        return Point<dim>(array,coordinate_system);
+      }
 
       /**
        * add two points
        */
-      Point<dim> operator+(const Point<dim> &point) const;
+      inline
+      Point<dim> operator+(const Point<dim> &point_right) const
+      {
+        WBAssert(coordinate_system == point_right.get_coordinate_system(),
+                 "Cannot add two points which represent different coordinate systems.");
+        Point<dim> point_tmp(point,coordinate_system);
+        for (unsigned int i = 0; i < dim; ++i)
+          point_tmp[i] += point_right[i];
+
+        return point_tmp;
+      }
 
 
       /**
        * Substract two points
        */
-      Point<dim> operator-(const Point<dim> &point) const;
-
+      inline
+      Point<dim> operator-(const Point<dim> &point_right) const
+      {
+        WBAssert(coordinate_system == point_right.get_coordinate_system(),
+                 "Cannot substract two points which represent different coordinate systems. Internal has type " << static_cast<int>(coordinate_system)
+                 << ", other point has type " << static_cast<int>(point_right.get_coordinate_system()));
```

Comment:
This is true for the dot product operator as well, by the way, even though it doesn't have this assertion.

Reply:
Moreover, the formula for the dot product is only correct for Cartesian coordinate systems :-)